src/server/utils/auth.openid.validator.js

```diff
@@ -0,0 +1,111 @@
+const _ = require('lodash');
+const request = require('superagent');
+const errors = require('./error.utils');
+
+/**
+ * Parse the `token` from the given `req`'s authorization header.
+ *
+ * @param {Request} req
+ * @return {String|null}
+ */
+function parseBearerToken(req) {
+
+	let auth;
+	if (!req.headers || !(auth = req.headers.authorization)) {
+		return null;
+	}
+
+	// split on space
+	const parts = auth.split(' ');
+	if (parts.length < 2) {
+		return null;
+	}
+
+	// get schema and token from array
+	const schema = parts.shift().toLowerCase();
+	const token = parts.join(' ');
+
+	// validate that it is a bearer
+	if (schema !== 'Bearer') {
+		return null;
+	}
+
+	return token;
+}
+
+/**
+ * Get valid scopes.
+ *
+ * @param {Object} scopes
+ * @param {Array<String>} allowedScopes
+ * @returns {Boolean}
+ */
+function getValidScopes(scopes, allowedScopes) {
+	const tokenScope = scopes;
+	const scopeArray = typeof tokenScope === 'string' ? tokenScope.split(/[, ]/) : [];
+	const intersection = _.intersection(scopeArray, allowedScopes);
+
+	return intersection;
+}
+
+/**
+ * @name validate
+ * @summary {Function} Returns a middleware function that verifies bearer token and checks scope
+ * @param {Array<String>} validScopes
+ * @param {Object} loggerUtil
+ * @param {Object} config
+ */
+module.exports.validate = (allowedScopes, logger, config) => {
+
+	return async (req, res, next) => {
+
+		// get bearer token
+		const bearerToken = parseBearerToken(req);
+
+		if (bearerToken) {
+			return request
+				.post(config.auth.introspectionUrl)
+				.set('content-type', 'application/x-www-form-urlencoded')
+				.send({token: bearerToken, client_id: config.auth.protectedResourceClientId, client_secret: config.auth.protectedResourceClientSecret})
+				.then((introspectionResponse) => {
+
+					const token = introspectionResponse.body;
+					if (token.active) {
+						const { scope, context, aud = '' } = token;
+
+						// make sure the token is intended for our server
+						if (aud && aud === config.auth.resourceServer) {
+							// get scopes that are valid for the endpoint specified
+							const scopes = getValidScopes(scope, allowedScopes);
+							// verify if token has one of the required scopes
+							if (scopes && scopes.length > 0) {
+								// attach patient id to req or validate request and patient
+								// TODO: verify this as it could be done in different ways
+								if (context && context.patient) {
+									req.patient = context.patient;
+								}
+								// if location
+								if (context && context.location) {
+									req.location = context.location;
+								}
+								// if encounter
+								if (context && context.encounter) {
+									req.encounter = context.encounter;
+								}
+								// validation complete
+								return next();
+							}
+						}
+					}
+					// default return unauthorized
+					return next(errors.unauthorized('Invalid token'));
```

Comment:
Can we add a second argument to the error utils?

I made an update in a recent PR that added a version parameter so the error utils can use a specific version of OperationOutcome.  It will default to stu3 for now. 

There is a version on all req objects via `req.params.version`.

Reply:
added version